File: ast/context.go
package ast

import (
	"fmt"

	"github.com/panda-io/micro-panda/ast/ast"
	"github.com/panda-io/micro-panda/ast/ast_types"
	"github.com/panda-io/micro-panda/ast/declaration"
)

func NewContext(p *Program) ast.Context {
	return &Context{
		Program: p,
		objects: make(map[string]ast.Type),
	}
}

type Context struct {
	Program  *Program
	Function *declaration.Function

	parent  *Context
	objects map[string]ast.Type
}

func (c *Context) Error(offset int, message string) {
	c.Program.Error(offset, message)
}

func (c *Context) NewContext() ast.Context {
	return &Context{
		Program:  c.Program,
		Function: c.Function,

		parent:  c,
		objects: make(map[string]ast.Type),
	}
}

func (c *Context) AddObject(name string, t ast.Type) error {
	if _, ok := c.objects[name]; ok {
		return fmt.Errorf("redeclared variable: %s", name)
	}
	c.objects[name] = t
	return nil
}

func (c *Context) FindObject(name string) ast.Type {
	if v, ok := c.objects[name]; ok {
		return v
	}
	if c.parent != nil {
		v := c.parent.FindObject(name)
		if v != nil {
			return v
		}
	}
	if c.Function != nil && c.Function.Parent != nil {
		return c.Function.Parent.MemberType(name)
	}
	return nil
}

func (c *Context) ResolveType(v ast.Type) ast.Type {
	switch t := v.(type) {
	case *ast_types.TypeName:
		d := c.FindDeclaration(t)
		if d == nil {
			c.Program.Error(v.GetPosition(), "type not defined")
		} else {
			t.Qualified = d.QualifiedName()
			switch i := d.(type) {
			case *declaration.Function:
				return i.Typ

			case *declaration.Struct:

			case *declaration.Enum:
				t.IsEnum = true

			default:
				c.Program.Error(v.GetPosition(), "type not defined")
			}
		}
		return t

	case *ast_types.TypeArray:
		t.ElementType = c.ResolveType(t.ElementType)
		if t.Dimension[0] < 0 {
			c.Program.Error(v.GetPosition(), "invalid array index")
		}
		for i := 1; i < len(t.Dimension); i++ {
			if t.Dimension[i] < 1 {
				c.Program.Error(v.GetPosition(), "invalid array index")
			}
		}
		return t

	case *ast_types.TypePointer:
		t.ElementType = c.ResolveType(t.ElementType)
		return t

	case *ast_types.TypeFunction:
		t.ReturnType = c.ResolveType(t.ReturnType)
		for i := 0; i < len(t.Parameters); i++ {
			t.Parameters[i] = c.ResolveType(t.Parameters[i])
			if ast_types.IsStruct(t.Parameters[i]) {
				c.Program.Error(t.Parameters[i].GetPosition(), "struct is not allowed as parameter, use pointer instead")
			}
			if ast_types.IsArray(t.Parameters[i]) {
				c.Program.Error(t.Parameters[i].GetPosition(), "array is not allowed as parameter, use pointer instead")
			}
		}
		return t

	default:
		return t
	}
}

func (c *Context) FindDeclaration(t ast.Type) ast.Declaration {
	typeName := t.(*ast_types.TypeName)
	if typeName.Qualified == "" {
		return c.FindLocalDeclaration(typeName.Name)
	}
	d := c.FindQualifiedDeclaration(typeName.Qualified)
	if _, ok := d.(*declaration.Enum); ok {
		typeName.IsEnum = true
	}
	typeName.Qualified = d.QualifiedName()
	return d
}

func (c *Context) FindLocalDeclaration(name string) ast.Declaration {
	qualified := fmt.Sprintf("%s.%s", c.Program.Module.Namespace, name)
	if d, ok := c.Program.Declarations[qualified]; ok {
		return d
	}
	qualified = fmt.Sprintf("%s.%s", ast.Global, name)
	if d, ok := c.Program.Declarations[qualified]; ok {
		return d
	}
	for _, i := range c.Program.Module.Imports {
		qualified = fmt.Sprintf("%s.%s", i.Namespace, name)
		if d, ok := c.Program.Declarations[qualified]; ok {
			return d
		}
	}
	return nil
}

func (c *Context) FindQualifiedDeclaration(qualified string) ast.Declaration {
	return c.Program.Declarations[qualified]
}

func (c *Context) IsNamespace(name string) bool {
	return c.Program.IsNamespace(name)
}

func (c *Context) SetFunction(f ast.Function) {
	c.Function = f.(*declaration.Function)
}

func (c *Context) GetFunction() ast.Function {
	return c.Function
}

File: app/src/main/java/com/github/panda_io/micro_panda/ast/Context.java
package com.github.panda_io.micro_panda.ast;

import java.util.*;

import com.github.panda_io.micro_panda.ast.type.*;
import com.github.panda_io.micro_panda.ast.declaration.*;
import com.github.panda_io.micro_panda.ast.declaration.Enumeration;

public class Context {
    Program  program;
    Function function;
	Context parent;
	Map<String, Type> objects;

    public int loopLevel;

    public Context(Program program) {
        this.program = program;
        this.objects = new HashMap<>();
        this.loopLevel = 0;
    }

    public Context newContext() {
        Context context = new Context(this.program);
        context.function = this.function;
        context.loopLevel = this.loopLevel;
        context.parent = this;
        return context;
    }

	public void addError(int offset, String message) {
        this.program.addError(offset, message);
    }

	public boolean insertObject(String name, Type type) {
        if (this.objects.containsKey(name)) {
            return false;
        }
        this.objects.put(name, type);
        return true;
    }

    public Type findObject(String name) {
        if (this.objects.containsKey(name)) {
            return this.objects.get(name);
        }
        if (this.parent != null) {
            Type object = this.parent.findObject(name);
            if (object != null) {
                return object;
            }
        }
        if (this.function != null && this.function.parent != null) {
            return this.function.parent.memberType(name);
        }
        return null;
    }

    public Type resolveType(Type type) {
        if (type instanceof TypeName) {
            Declaration declaration = this.findDeclaration(type);
            if (declaration == null) {
                this.addError(type.getOffset(), "type not defined");
            } else {
                if (declaration instanceof Function) {
                    return ((Function)declaration).type;
                } else if (declaration instanceof Struct) {
                } else if (declaration instanceof Enumeration) {
                    ((TypeName)type).isEnum = true;
                } else {
                    this.addError(type.getOffset(), "type not defined");
                }
            }
        } else if (type instanceof TypeArray) {
            TypeArray array = (TypeArray)type;
            array.elementType = this.resolveType(array.elementType);
            if (array.dimensions.size() == 0 || array.dimensions.get(0) < 0) {
                this.addError(type.getOffset(), "invalid array index");
            }
            for (int dimension : array.dimensions) {
                if (dimension < 1) {
                    this.addError(type.getOffset(), "invalid array index");
                }
            }
        } else if (type instanceof TypePointer) {
            TypePointer pointer = (TypePointer)type;
            pointer.elementType = this.resolveType(pointer.elementType);
        } else if (type instanceof TypeFunction) {
            TypeFunction function = (TypeFunction)type;
            function.returnType = this.resolveType(function.returnType);
            for (int i = 0; i < function.parameters.size(); i++) {
                Type parameter = this.resolveType(function.parameters.get(i));
                function.parameters.set(i, parameter);
                if (parameter.isStruct()) {
                    this.addError(parameter.getOffset(), "struct is not allowed as parameter, use pointer instead");
                } else if (parameter.isArrayWithSize()) {
                    this.addError(parameter.getOffset(), "array is not allowed as parameter, use pointer instead");
                }
            }
        }
        return type;
    }

	public Declaration findDeclaration(Type type) {
        TypeName name = (TypeName)type;
        if (name.qualified == null) {
            return this.findLocalDeclaration(name.name);
        }
        Declaration declaration = this.findQualifiedDeclaration(name.qualified);
        if (declaration instanceof Enumeration) {
            name.isEnum = true;
        }
        name.qualified = declaration.qualified;
        return declaration;
    }

	public Declaration findLocalDeclaration(String name) {
        String qualified = String.format("%s.%s", this.program.module.namespace, name);
        if (this.program.declarations.containsKey(qualified)) {
            return this.program.declarations.get(qualified);
        }
        qualified = String.format("%s.%s", Constant.global, name);
        if (this.program.declarations.containsKey(qualified)) {
            return this.program.declarations.get(qualified);
        }
        for (Module.Using imported : this.program.module.usings) {
            qualified = String.format("%s.%s", imported.namespace, name);
            if (this.program.declarations.containsKey(qualified)) {
                return this.program.declarations.get(qualified);
            }
        }
        return null;
    }

	public Declaration findQualifiedDeclaration(String qualified) {
    	return this.program.declarations.get(qualified);
    }

	public boolean isNamespace(String ns) {
        return this.program.isNamespace(ns);
    }

    public void setFunction(Function function) {
        this.function = function;
    }

    public Function getFunction() {
        return this.function;
    }
}
